arweave/caip2.md

```diff
@@ -0,0 +1,56 @@
+---
+namespace-identifier: arweave-caip2
+title: Arweave Namespace - Chains
+author: Rohit Pathare (@ropats16), Phil Billingsby (@pbillingsby)
+discussions-to:
+status: Draft
+type: Standard
+created: 2022-09-01
+requires: CAIP-2
+---
+
+# CAIP-2
+
+*For context, see the [CAIP-2][] specification.*
+
+## Abstract
+In CAIP-2 a general blockchain identification scheme is defined. This is the implementation of CAIP-2 for Arweave network.
+
+### Arweave Namespace
+
+The namespace "arweave" refers to the wider Arweave ecosystem.
+
+#### Reference Definition
+
+The reference relies on Arweave's current designation of addresses belonging to main networks by prefixing them with `1`.
+
+
+## Rationale
+
+Arweave has one blockchain, the [Blockweave](https://www.arweave.org/technology#blockweaves).
+
+## Backwards Compatibility
+
+Not applicable.
+
+## Test Cases
+
+This is a manually composed example.
+
+```
+# Arweave Mainnet
+arweave:1
```

Comment:
Arweave has only 1 chain currently which has been designated `1` in the ecosystem. If more chains are added, a system for identifying chains would be created at the time and the `arweave` namespace would be updated accordingly.

Reply:
There is no testnet at all? Can you also include a reference to where it's designated `1` in the ecosystem?

Reply:
There is no official test net yet. The designation of the main network as 1 is a value referenced from the EVM compatible ecosystem. As this kind of standard was not needed before this particular use case, there is no document I can point to for the same. This is essentially the first iteration of such a standard for the arweave ecosystem.

Reply:
In ethereum the number has a specific meaning and all network id's are maintained on a list: https://chainlist.org/

Are there unofficial test nets? If yes, how does the client software differentiate between them?

Reply:
The Arweave protocol doesn't have built in support for multiple networks at the protocol level. HTTPS request are made to nodes but there's a single official Arweave network so there isn't a need to get a response from the nodes as to which network they are on. In case this is revised in the future, adequate documentation and PRs for the same would be submitted.

Reply:
From a quick google it seems like there are indeed testnets. On a protocol level they need to be distinguished. Also if we arbitrarily pick a number here it will fall on CASA to maintain a list of numbers for arweave chains which we definitely don't want to do!

Reply:
There are testnets but they aren't actually networks of nodes so there's no need to distinguish them at a the protocol level. They aren't gossping data/transactions  on an open network.  These "testnets" expose Arweave like HTTP endpoints but are hosted as a standalone deployments.  When a developer wants to test their application they point their code at  the URL provided by the testnet and communicate directly with it. Testing works like this on arweave, in part, because the arweave protocol does not have a way to differentiate "network" at the protocol level. 

Development on Arweave is typically done using https://github.com/textury/arlocal, a shim that mocks up the arweave interfaces and mimics its behavior. The "testnets" that exist are a slightly modified version of this code that enable incoming connections to be managed by a load balancer. Example of this would be Redstone's "testnet" https://testnet.redstone.tools/

You may also find references to Testweave-sdk or Testweave-docker. Testweave coexisted with `arlocal` for some time and similarly enabled testing on the developers local machine. Instead of a node/SQLite mockup testweave ran an Arweave node in a container. Arweave node software really wasn't meant to run in this configuration and as a result it had constant bugs and inconsistencies with mainnet. Developers abandoned testweave for the better supported `arlocal` and it is now effectively deprecated with no active maintainer.

Reply:
is there any canonical "chainID" equivalent for these testnets, though, canonically maintained in any kind of registry on github or otherwise, as Joel was asking? would "testnet.redstone.tools" be a meaningful testnet name, on which someone might want to test a system by addressing and fetching state?

this information looks pretty useful to first-time arweave devs, would it make sense to put it into the README somewhere as an informational aside?

Reply:
@DanMacDonald Arweave is releasing a testnet for 2.6. Isn't this going to be its own chain?

Reply:
That's true, there is a testnet for 2.6, purely for miners. I'm not sure exactly how they are accomplishing this. My guess is though a whitelist of 2.6 peers.

Reply:
Hey @oed @bumblefudge, based on the information provided by @DanMacDonald, how can we move forward? As the testnet for 2.6 isn't released yet, ans it's working mechanism is still vague, is it possible to move forth with considering the existence of only the Arweave main net at the moment and once the test net is released, the chain resolution can be revised and updated accordingly in the namespaces as well?

Reply:
Can we maybe use the genesis has as we do for other chains? That way if in the future Arweave introduces a testnet we don't have to maintain a list of numbers.

Reply:
this makes sense to me-- calling it `1` or `main` is arbitrary if the protocol doesn't internally have the concept of a chain ID and the community itself doesn't use that convention.  i worry that, as Joel rightly points out, the community might proceed happily for years without testnets or forks, and at such a time as they adopt the concept, they might number from 0 or use strings and call it `main-net` or `mainnet` instead of `main`.  For this reason, I second the recommendation for the "hash of genesis block" approach taken by many other caip-2 profiles since current protocol does seem to have the notion of "block 0" and addressing blocks by hashes.  

from [here](https://viewblock.io/arweave/block/0) I take it that the genesis hash is numbered 0, not 1, and hash-addressable as `7wIU7KolICAjClMlcZ38LZzshhI7xGkm2tDCJR7Wvhe3ESUo2-Z4-y0x1uaglRJE` 
. from [this documentation] i am inferring it that since block 422250 the format of RPC calls changed, but the block hashes still seem at first glance to be computed the same way and would fit the same regex: 
`JbtpfyJ6Kh22vciZJpZAVorejqDRB6ekbZC0QF_zOFbACpcae6L6PZsuy0gKRpqk` . 
It's interesting that the RPC doesn't include a request for block by block number, just by hash, but at least you can confirm you're on the intended chain (or at least, a fork of it...) by calling 
[`GET https://arweave.net/block/hash/7wIU7KolICAjClMlcZ38LZzshhI7xGkm2tDCJR7Wvhe3ESUo2-Z4-y0x1uaglRJE`](https://docs.arweave.org/developers/server/http-api#blocks) and checking the `indep_hash` top-level property of the result object to confirm.  It seems like that flow still works after the info RPC changes.

Anyways if this works for everyone, feel free to update the CAIP-2 profile and we can go from there-- tag me i you'd like any help.  To limit the length of CAIP2 and CAIP10 URNs to fit inside, say, HTTP headers, we have historically just substringed the first 16 or 32 characters-- the full 64 would make for unworkably long CAIP-10s and CAIP-19s.

Reply:
A trunkated hash of genesis might be alright as well. So the caip2 doesn't become so long. Maybe first 4 characters?

Reply:
Hey @bumblefudge @oed, just updated the PR to identify the network using a truncated version (first 16 characters) of the [hash of the genesis block](https://viewblock.io/arweave/block/0) in both `CAIP2` and `CAIP10` files.